### ISSUE TITLE
elasticsearch_xpack_role  missing allow_restricted_indices parameter

### ISSUE THREAD

**Issue by wasilak** [phillbaker/terraform-provider-elasticsearch#282]
Hi,

It'll be awesome if you could implement support for `allow_restricted_indices` param in `elasticsearch_xpack_role`
as described here: https://www.elastic.co/guide/en/elasticsearch/reference/current/defining-roles.html

Thanks! :)

**Comment by phillbaker** [comment]
Hello, happy to review a PR that adds this. It can be added to the schema here: https://github.com/phillbaker/terraform-provider-elasticsearch/blob/1e8c6d082bfa31fd8804aefbf49ca24df70933d3/es/resource_elasticsearch_xpack_role.go#L31 and then threaded through the relevant objects and expand/flatten functions.

**Comment by wasilak** [comment]
Hey @phillbaker

I tried and almost succeeded. Almost :) It looks like there will be larger refactoring required in order to support this param as it needs Elasticsearch v8 and as far as I can tell your provider relies on `https://github.com/olivere/elastic` which does not support v8 (yet?) which means that

```golang
import (
[...]
elastic8 "github.com/olivere/elastic/v8"
)
```
won't work.

Trying to use this param with v6 or v7 results in:
```
# github.com/phillbaker/terraform-provider-elasticsearch/es
es/util.go:273:34: item.AllowRestrictedIndices undefined (type "gopkg.in/olivere/elastic.v6".XPackSecurityIndicesPermissions has no field or method AllowRestrictedIndices)
es/util.go:281:34: item.AllowRestrictedIndices undefined (type "gopkg.in/olivere/elastic.v6".XPackSecurityIndicesPermissions has no field or method AllowRestrictedIndices)
es/util.go:299:34: item.AllowRestrictedIndices undefined (type "github.com/olivere/elastic/v7".XPackSecurityIndicesPermissions has no field or method AllowRestrictedIndices)
es/util.go:307:34: item.AllowRestrictedIndices undefined (type "github.com/olivere/elastic/v7".XPackSecurityIndicesPermissions has no field or method AllowRestrictedIndices)
```

Fortunately as far as I can tell this particular param can be added via Elasticsearch API call and it won't be affected by current version of your provider :)

Something like that:
```
PUT /_security/role/your_role
{
  "indices": [
    {
      "names": [ "index-name-*" ],
      "privileges": ["all"],
      "allow_restricted_indices": true
    }
  ]
}
```
I guess, at least for now, we can park this topic and use workaround.

Thanks!

**Comment by phillbaker** [comment]
It's supported in v7 - no need to upgrade to v8, see the v7 docs here: https://www.elastic.co/guide/en/elasticsearch/reference/7.17/defining-roles.html.

It would need to be added into the upstream library being used in this repo, the struct here would need to be updated: https://github.com/olivere/elastic/blob/29ee98974cf1984dfecf53ef772d721fb97cb0b9/xpack_security_get_role.go#L181-L186

**Comment by wasilak** [comment]
Yep. it was not implemented in https://github.com/olivere/elastic/blob/release-branch.v7/xpack_security_get_role.go#L181

```go
// XPackSecurityIndicesPermissions is the indices permission object
type XPackSecurityIndicesPermissions struct {
	Names         []string    `json:"names"`
	Privileges    []string    `json:"privileges"`
	FieldSecurity interface{} `json:"field_security,omitempty"`
	Query         string      `json:"query"`
}```

That is why I get error when adding it to `flattenIndicesPermissionSetv7()` in `util.go`:
```
es/util.go:305:34: item.AllowRestrictedIndices undefined (type "github.com/olivere/elastic/v7".XPackSecurityIndicesPermissions has no field or method AllowRestrictedIndices)
```

I'll try to submit PR to upstream repo :)
